KIPs/kip-81.md

```diff
@@ -0,0 +1,335 @@
+    // ...
+    _to.transfer(_value);
+    stakingTracker.notifyStake();
+    // ...
+  }
+  function withdrawApprovedStaking(uint256 _approvedWithdrawalId) {
+    // ...
+    _to.transfer(_value);
+    stakingTracker.notifyStake();
+    // ...
+  }
+}
+```
+
+In addition to balance notification, CnStakingV2 can appoint an address as the GC's voter account. The voter account can vote on behalf of this GC's all voting powers. The appointing happens through the existing multisig facility.
+
+```solidity
+contract CnStakingV2 {
+
+  function submitAppointVoter(address _voter) external
+  afterInit()
+  onlyAdmin(msg.sender) {
+    uint256 id = requestCount;
+    submitRequest(id, Functions.AppointVoter, bytes32(_voter), 0, 0);
+    confirmRequest(id, Functions.AppointVoter, bytes32(_voter), 0, 0);
+  }
+  function appointVoter(address _voter) external
+  onlyMultisigTx() {
+    stakingTracker.appointVoter(_voter);
+  }
+}
+```
+The new CnStakingV2 contract will have version number 2 so other contracts can distinguish from the old version.
+```solidity
+contract CnStakingV2 {
+  uint256 constant public VERSION = 2;
+}
+```
+
+#### StakingTracker contract
+
+Each GC stake its KLAYs in its own contracts. Even one GC is allowed to have multiple CnStaking contracts, in which case its stake amount must be added up. Because stake amounts are scattered across multiple contract accounts, it is not trivial to calculate their voting powers.
+
+The StakingTracker contract is introduced to record GC stake amounts at a central location. To host a voting, following data are needed.
+
+- Number of eligible GCs: to calculate the maximum votes, and the turnout quorum condition.
+- Total eligible votes: to calculate the turnout quorum.
+- Per-GC votes: to process voting requests (i.e. calls to Voting contract).
+- Per-GC voting accounts: to process voting requests (i.e. calls to Voting contract).
+
+The data are collected on a per-proposal basis. For each proposal, a Tracker data structure is created and updated until a specified block.
+
+```solidity
+interface IStakingTracker {
+  struct Tracker {
+    // this struct is updated only if trackStart <= block.number < trackEnd.
+    uint64 trackStart; // when this Tracker is created
+    uint64 trackEnd;   // specified by createTracker(), voting startBlock.
+
+    // List of eligible nodes and their staking contracts.
+    // Determined at createTracker() and won’t change.
+    address[] nodeIds;
+    mapping(address => address) stakingToNodeId;
+
+    // Balances and voting powers of each nodes.
+    // Updated by notifyStake().
+    mapping(address => uint256) stakingBalances;
+    mapping(address => uint256) nodeBalances;
+    mapping(address => uint256) votingPowers;
+    uint256 totalVotes;
+
+    // Appointed voters list
+    // Updated by appointVoter().
+    mapping(address => address) nodeIdToVoter;
+    mapping(address => address) voterToNodeId;
+  }
```

Comment:
It would be good if we define prototypes of the functions in this contracts.

Reply:
The StakingTracker interface and example implementations are given in the updated document (https://github.com/yeri-lee/kips/blob/master/KIPs/kip-81.md#stakingtracker)